Issue 6451: ContainerKeyIndex may be throttling updates to system critical segments

RaulGracia opened the pull request:
**Change log description**  
Adds a new property for KVTables named "systemcritical.unindexed.bytes.max" that defines the amount of unindexed data we allow to system-critical KVTable Segments (defaulted to 256MB, 2x the amount of unindexed data we allow to regular segments).

**Purpose of the change**  
Fixes #6451 .

**What the code does**  
In some cases, when LTS is faulty or unavailable for some time, a single Segment Container can accumulate around 1M Operations before the throttler starts inducing max delays. However, we have seen that, in practice, this can lead to an accumulation of a lot of updates on KVTable Segments, and specially on system-critical ones. If the system gets to this state, it may happen that the `StorageWriter` cannot start upon a container recovery, as during its initialization it may be necessary to perform updates on some system-critical KVTables. 

This PR essentially does two things to mitigate this problem:
- First, in a similar spirit to what the ingestion path does by not throttling system-critical Segments, we provide by default a higher (2x) throttling threshold (i.e., unindexed data length) to system-critical KVTable Segments. This should prevent up to some extent to get into the issue that motivates this PR.
- Second, the unindexed data length that we allow to system-critical Segments is now defined by a configuration parameter, namely `systemcritical.unindexed.bytes.max`. This will allow us, in the case we find the same issue, to increase the credit only in the relevant KVTable Segments, instead on the whole system (which may lead to high memory pressure).

**How to verify it**  
Added new tests to validate that system-critical and regular segments get the right amount of credits.

andreipaduroiu commented on segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java:
```diff
@@ -842,13 +847,15 @@ void updateSegmentIndexOffset(long segmentId, long indexOffset, int processedSiz
                     val si = segment.getInfo();
                     long initialDelta = Math.max(0, si.getLength() - IndexReader.getLastIndexedOffset(si));
-                    throttler = new AsyncSemaphore(config.getMaxUnindexedLength(), initialDelta, String.format("%s-%s", containerId, segment.getSegmentId()));
+                    throttler = new AsyncSemaphore(totalCredits, initialDelta, String.format("%s-%s", containerId, segment.getSegmentId()));
```
Make sure that when you release the throttle, you also account for this case. 

Also let's add some unit tests too so that we're covered here.

sachin-j-joshi commented:
The current change basically postpones the problem by increasing the limit, which is fine but there are two things
- major issue is - we'll still have the same deadlock sometime somewhere.
- minor issue is - What is a new a safe limit ? is 8 times sufficient ?

Also, 
We need to log WARN/ERROR when system critical segments are actually throttled. This is extremely important to know, as the end result is a deadlock and halting of the container.
It would also be helpful to know size of indexed pending data.

The other problem - if we never run into this problem and go into deadlock then how do we get out of it? Perhaps changing config value is the solution here but we need to document it somewhere.

sachin-j-joshi reviewed (changes requested):
See above comment.

realAaronWu commented:
> The current change basically postpones the problem by increasing the limit, which is fine but there are two things
> 
> * major issue is - we'll still have the same deadlock sometime somewhere.
> * minor issue is - What is a new a safe limit ? is 8 times sufficient ?
> 
> Also, We need to log WARN/ERROR when system critical segments are actually throttled. This is extremely important to know, as the end result is a deadlock and halting of the container. It would also be helpful to know size of unindexed pending data.
> 
> The other problem - if we ever run into this problem and go into deadlock then how do we get out of it? Perhaps changing config value is the solution here but we need to document it somewhere.

Can we force apply changes for system critical segments instead of increasing credits? If the side effect is just lengthy delays upon a subsequent recovery?

RaulGracia commented:
@sachin-j-joshi @realAaronWu  I have addressed the comment of the warning: now if we have a system-critical table segment that runs out of credit, it will be logged with a warning.

I want to be clear regarding the approach of this PR: it may not be completely solving the deadlock, because the deadlock we observe is in itself an unfortunate corner case of the design we have involving (hash-based) Table Segments + SLTS. As you know, there is an inherent circular dependency in hash-based Table Segments, as the updates go through the ingestion pipeline and are stored in tier 1 (and cached), but they don't actually make effect until the Storage Writer workflow makes the changes in the underlying BTree index backed on storage. If the ingestion pipeline makes progress but Tier 2 is unavailable, the Storage Writer won't be able to make progress and unindexed data will be accumulated. On top of that, we also have SLTS which also depends on performing updates (not only reads) to Table Segments for initializing (another circular dependency), and these updates may be related to Table Segments that have accumulated a lot of unindexed data, so they cannot accept any further update until the amount of unindexed data is reduced, which is not going to happen.

This PR just does 2 things:
- It distinguishes the amount of credit (i.e., unindexed data length) that we allow to regular or system-critical Segments. In the case we find the same problem again, updating the new configuration value will unblock the cluster inducing minimal memory pressure (as the configuration change would only impact system-critical segments instead of all of them). Before, in similar situations we have increased credit on all Table Segments.
- The new default allows system-critical Table Segments to accumulate up to 2x the amount of data than regular ones. This is expected to reduce the chances that this happens again. However, we do not know the upper bound in terms of unindexed data for a hash-based Table Segment, which implies that we don't know if this could happen again. We need an empirical evaluation of that (and perhaps new metrics) to understand this in depth. I have created a new issue to track this activity, which can help us to be sure whether this change completely removes the problem or not: #6467.

Note that in the case that as part of #6467 we set a right default that i) is the upper bound of the maximum unindexed data that a system-critical Table Segment can accumulate, ii) we evaluate that the memory pressure of this new default is not going to be a problem during recoveries, then I would say that the problem would be definitively solved. In a similar spirit than the circular dependency of SLTS metadata with Tier 1 was solved by adding a throttling exception in the write path for system-critical Segments.

sachin-j-joshi reviewed (approved):
LGTM.

Let's run system tests and other sanity tests before merging.

RaulGracia commented:
@sachin-j-joshi system tests have passed (`test-pravega-continuous-1/1864/`)
